### ISSUE TITLE
Orlando's Billing Master Cycle *URGENT*

### ISSUE THREAD

**Issue by kdjstudios** [kdjstudios/SABillingGitlab#1739]
In GitLab by @kdjstudios on Oct 10, 2019, 10:54

**Submitted by:** "Kimberly Gagner" <kimberly.gagner@answernet.com>

**Helpdesk:** http://www.servicedesk.answernet.com/profiles/ticket/9490200

**Server:** Internal

**Client/Site:** Orlando

**Account:** NA

**Issue:** 

I am reviewing SAB reports SAB SBS shows the base rate billed for the entire cycle as $95.00. 

 

I have rerun a billing export and reverted and uploaded but it is still incorrect.  

 

Please call my cell as soon as possible 603-978-1555 as we need to get this corrected before the EOD.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 10, 2019, 11:42

@kdjstudios @jredfield 

Thank you for the update.

* It would be great if you provide the  Billing Cycle details and which report you are talking about.
* I am a little bit confused about the report `SBS`  . Does this means `SummaryBillingJournal`

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 10, 2019, 11:58

@amishra 

SBS - Service Code Breakdown

The latest Master billing cycle. EDIT: it looks like ops may have reverted the billing cycle now, as it is now showing at the upload step with no invoices in draft or finalized.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 10, 2019, 12:02

I wrote:

> The support team noticed that the last master billing cycle for Orlando (09/28/2019 Master) is now at the upload step, which means there are no invoices yet to be created. Did someone perform a revert?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 10, 2019, 15:06

David wrote:

> Export has been uploaded again.

I wrote:

> Please continue to process your billing and let us know if you encounter any issues.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 10, 2019, 16:16

@smasih @amishra 

The site was able to upload and process their billing cycle. It is now back at showing the same issue before. Please search why the base rates are not being invoiced and provide an estimate to fix the full issue (not just a band-aid one time fix). This needs to be completely fixed by tomorrow morning so the site can complete their billing.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 10, 2019, 16:23

Kim mentioned via IM, it looks like the "Service Base Rate - 33050" is the code that is not being invoiced.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 11, 2019, 05:46

@kdjstudios @jredfield 

Thank you for the update.

* There is another ticket #1560  related to the same issue. Sumeet's have explained in detail on comment http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1560#note_47055 and Tim's comment http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1560#note_47235 as well.  Please go through that ticket and see Sumeet's & Tim's conversations.

* We looked one account(Account Number `045-W01087`) of `Orlando, FL` site.

* Account has one active additional fee/recurring fee and this fee is not reflecting on the current invoice because the Billing cycle's service start date month is **9** and last accessed date is **9** but, turn on date month is **10** which cause the issue.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 11, 2019, 06:24

assigned to @amishra

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 11, 2019, 08:07

@kdjstudios @jredfield 

As per today's huddle, 

Tim brought up that #1560 is the change due to which the recurring fees were not billed **but, on revert the "last assessed date" should have been updated**

There could be a possible code/data issue which is stopping "last assessed date" on revert.

Currently we ran a manual query *(to set back access date)* on **live server for this site**  so that Orlando can bill today. (we did this after running our test on merge server)  

List of the updated accounts is attached here. [update_assessed_date_for_site_orlando.csv](/uploads/172793a27223ca4f3fda4e3ceb2e3db5/update_assessed_date_for_site_orlando.csv)

Below is described which step needs to perform.

* Revert the master Billing Cycle for `Orlando Site`.
* Upload the Usage file or run as blank.
* Check some random account draft invoices and you are able to see the recurring fee/additional fee.


#### We will continue to find the core reason for the issue tomorrow.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @jredfield on Oct 11, 2019, 08:14

@amishra @kdjstudios Ashish, they need the billing done by end of today--very urgent. So you are saying that with the manual query that sets back the access date, they can bill now?

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 11, 2019, 08:38

@kdjstudios @jredfield 

Thank you for the update.

Yes, they can bill now.

Below is described which step needs to perform.

* Revert the master Billing Cycle for `Orlando Site`.
* Upload the Usage file or run as blank.
* Check some random account draft invoices and you are able to see the recurring fee/additional fee.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 11, 2019, 08:51

I wrote:

> The support team has finished performing their last checks and the site will need to Revert the billing cycle and reprocess. Please let us know if the Base rates populate on the draft invoices. Thank you and we look forward to your response.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 11, 2019, 08:54

@amishra @smasih @jredfield 

So the issue is that when sites are performing a revert, the last assessed date is not properly being reset on fees?
 Or is this the issue mentioned in #1560 where someone changed a fee at the site level while processing a billing cycle? Or a combination of both? I still do not fully understand what SAB is failing at doing.

**Comment by kdjstudios** [comment]
In GitLab by @jredfield on Oct 14, 2019, 05:36

Kim is still waiting for an explanation of what caused this issue. Please investigate further.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 14, 2019, 07:19

@kdjstudios @jredfield 


Thank you for the update.

After deeply researched the issue, We found that if the first Billing cycle service period start date `Month & Year` is the same service period start date `Month & Year` of the next Billing Cycle then the system ignores the additional fee/recurring fee. 

@tim.traylor @jredfield 

`````ruby
def applicable_after_last_assessed_fee_check?(last_assessed_date, billing_date, account_id)
    account = Account.find(account_id)
    result = false
    if last_assessed_date.month == billing_date.month && last_assessed_date.year == billing_date.year
      if account.turn_on_date.month == billing_date.month && account.turn_on_date.year == billing_date.year
        result = true
      end
    else # we did not already assess the fee this month
      result = true
    end
    result
  end
````` 

As per the current functionality of SAB, If the service start date(*Month & Year*) is the same with the assessed date(*Month & Year*) of the additional fee then we check account's **turn on date** (*Month & Year*) is same as billing cycle's service start date(month & year). If both are same then the system applies the additional fee otherwise, it ignores.

![image](/uploads/1afadc242b2bdd586283041a4babb4f1/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @jredfield on Oct 14, 2019, 08:26

@kdjstudios @amishra 

Ashish wrote:

> As per the current functionality of SAB, If the service start date(_Month & Year_) is the same with the assessed date(_Month & Year_) of the additional fee then we check account's **turn on date** (_Month & Year_) is same as billing cycle's service start date(month & year). If both are same then the system applies the additional fee otherwise, it ignores.

The code you are referring to does behave as you said.  However, it also appears that if the last assessed date (month/year) is *not* the same as the service start date, then the additional fee is not applied, and the check for the turn on date is not even necessary.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 14, 2019, 09:04

Kim wrote:

> Everything looks good the base rates are now showing.  David is in the final stages of is billing. 
> 
> Can you tell me what caused this issue?

I wrote:

> We are glad to hear the site was able to continue processing the billing cycle. We are will awaiting a clear response from the support team on what the exact cause was. Once we have it, we will be sure to provide you with an explanation. 


@amishra 

Thanks for the update above, however I am still uncertain on what the exact cause of this was. May you please review this with @jredfield tomorrow so he may follow up with Kim on the explanation.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Oct 15, 2019, 09:24

@jredfield @kdjstudios

Not sure if this is a bug, in short if the Service start of current and previous billing cycle are in same month then in that case system will filter out the recurring fees so that recurring is only charged once in a month on invoice.

As we can see in the screenshot the service start for the master was in the same month (september)

![image](http://gitlab.aavaz.biz/AnswerNet/SABilling/uploads/1afadc242b2bdd586283041a4babb4f1/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 15, 2019, 09:33

@smasih @jredfield 

Yes this is an bug as this should only be charging a 'one change per a month' if it is marked as part of a Billing Period. Thus to prevent the double charging on those accounts. Regular account setup should not take into consideration this check for double billing in the same month.

Would this be as simple as adding a check to the existing "one charge per month" functionality to see if it is eligible by checking if it is associated with a billing period? If so then do the check, if not, then process normally?

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 16, 2019, 02:11

@kdjstudios @jredfield 

We will do as instructed, but just so you know that these specs were laid out by Tim [here](http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1037#note_26648) 
It could be possible that we are not anticipating all the scenarios.

I will wait till the final specs are defined.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Oct 16, 2019, 05:39

@amishra @kdjstudios @jredfield hi Guys - this is definitely a flaw in the design for sites on a 28 day billing cycle. Originally I think we were looking at quarterly fees, so that may have been part of the design issue.  John, can you please work with Kyle and try and come up with a design that addresses this?  Then we can review.  Thx!

**Comment by kdjstudios** [comment]
In GitLab by @jredfield on Oct 16, 2019, 05:45

@tim.traylor @kdjstudios @amishra Yes we will work on a design to resolve this properly.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 16, 2019, 08:46

@jredfield 

I believe I already came up with a solution in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1601#note_49354. Please review and then put back for development.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 30, 2019, 14:48

@amishra @smasih @jredfield 

Looks like we are having more sites reporting this issue every billing cycle. This is an urgent ticket that needs to have the issue resolved.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Nov 4, 2019, 24:02

@kdjstudios 

Please update the priority, or is it still urgent.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 4, 2019, 24:40

@kdjstudios @jredfield 

*Please approve the estimate so that we'll start the development.*

## Developer Estimate
|Size|Lower Time Bound|Variability|Upper Time Bound|Testing|Time for Testing| Time for Dev & Testing|
|-|-|-|-|-|-|-|
|R-Small|2|Concepts and Implementation Well Understood|2.5|High Impact lots of use cases|2.5|5|

##### Estimate Notes:
* Risk areas identified: `Account/Invoice`



Thanks

**Comment by kdjstudios** [comment]
In GitLab by @jredfield on Nov 4, 2019, 05:42

Approved.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 4, 2019, 08:59

@jredfield @amishra @smasih 

I also just want to notate that we will need to have this fix released ASAP to try and prevent more billing cycle issues.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 5, 2019, 06:53

@kdjstudios @jredfield 

Thank you for the update.

We have completed the development and the testing part is left. I'll update you once setup branch for your testing.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 6, 2019, 03:19

@kdjstudios @jredfield 

Thank you for the update.

As Kyle's mentioned above on [comment](http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1601#note_49354), use billing period if exist then mark `one charge per month` as `true` otherwise process normally.

But, the billing period only checks month's numeric value, that's why this will not work.

We have added an additional query in existing functionality,

* If the fee is already applied on the account for the current billing cycle then work normally otherwise we add the fee to the invoice.

* This will work on every known case.

**We have set up a branch on the merge2 server Please test it.**

For testing this scenario you can use the Billing Cycle of `Triad`, and `Stockton` site.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Nov 6, 2019, 06:10

@kdjstudios @amishra @smasih 

My concern is that the new 'fix' that allows monthly fees to be assessed properly on 28 days cycles may also break the code that prevents multiple billing of quarterly and annual fees.

* For example, we have an account that has a quarterly fee, billed on the months 1, 4, 7, 10.
* The account is in a site that has a 28 day billing cycle.
* They have a cycle on 4/1 and we assess the quarterly fee.
* The site has another cycle on 4/29, and with the new code we check that the fee was not assessed in this same billing cycle.  It was not, so we allow the fee to be assessed even though we already assessed it for April.


My original recollection was that we were to look at the assessment months for annual and quarterly fees and compare that month to the last assessed date for the account. If the month matched, then we knew that we had already made the assessment for the intended period.

When we introduced monthly fees, this broke what we had because it can't work with the 28 day cycles. 

Questions:
* Why did we add monthly fees to this process?  
* When an account has had a base rate assessed by a manual invoice in the same billing cycle, aren't we supposed to skip that fee if the account was just turned on?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 6, 2019, 14:04

Lets discuss this during tomorrows meeting.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 7, 2019, 09:27

This was discussed on today's call, it was clarified we wanted anyone to voice any reason why we would want this check on regular recurring fees (non-Billing Period). No one could think of any reason why. I also found the following tickets and after a quick review of each could also not find a reason why we would. #1037 #1098 

If anyone does think of a reason please post it here, so we may update this fix will still accommodate that scenario.

Tim had mentioned about the turn on date for accounts and it should be taken into account. I was able to find that in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1037#note_26648. That would mean what I thought we reviewed with Gary was incorrect bout them having to manually remove it. But it should already be in the code to accommodate for this. Please make sure with the existing fix it still functions as expected.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Nov 11, 2019, 02:45

@jredfield 

Hi John, we are confused and not sure if we should continue with the development/fix as explained by Tim [here](http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1601#note_50687)

@kdjstudios 
Not sure if we need to keep the monthly recurring on manual invoice and give ops capability to delete it **or not?**

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 11, 2019, 11:02

@smasih @jredfield 

1. We need to only be checking for "duplicate" charges on fees that are marked as part of a billing period. All normal (non-billing period) fees should not be checked.
2. We should not be changing the functionality on manual invoicing, that should behave just as it is setup, for now. This we may need to discuss with Gary for his input.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 14, 2019, 04:07

@kdjstudios @jredfield 

Thank you for the update.

* We have developed the functionality as requested @tim.traylor  [Here](http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1601#note_50687) for the Billing period(*If billing period is present on the Fee*) and testing is left. We will let you know once ready for your testing.


* As per the current functionality for (**NON-BILLING PERIOD FEE**) If the fee is already applied on the account for the current billing cycle then work normally otherwise we add the fee to the invoice.**In short, the fee will not be applied twice if already apply for the current billing cycle.** `Do we need to change this functionality Please confirm with Gary?`


Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 14, 2019, 10:17

@amishra 

In regards to the second, this is correct. "The fee will not be applied twice if already apply for the current billing cycle."

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Nov 15, 2019, 04:19

@kdjstudios @jredfield 

Thank you for the prompt answer,

We have set-up a branch on the merge2 server for your testing.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 15, 2019, 11:12

I believe I was able to test this correctly, and it does appear the recurring fees appeared on both invoices in the same calendar month when there is no Billing period selected. I then tested with a billing period selected and it only appeared on the first invoice.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 15, 2019, 11:14

@jredfield 

With our next release not until the 23rd. I think we may want to do an ER release on this, I just want to ensure no more billing cycles have this issue. What do you think?

**Comment by kdjstudios** [comment]
In GitLab by @jredfield on Nov 15, 2019, 13:15

@kdjstudios Agreed. Tomorrow (11/16) is probably too soon, but maybe early next week.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 15, 2019, 13:44

Thanks @jredfield I will get it queued up. 

@amishra @ranand 

May we please have this released on Monday or Tuesday?

NOTE: I am going to also keep the announcement label on this as I want to ensure it goes out on the newsletter too as a bug fix.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 18, 2019, 01:08

@kdjstudios @jredfield 

Thank you for the update.

We will release this tomorrow(Tuesday) morning.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @jredfield on Nov 18, 2019, 06:25

@kdjstudios Kyle, they said they would release this tomorrow (Tuesday) once you confirm the announcement of this release.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 18, 2019, 11:40

@jredfield 

Thanks, Please release this bug fix.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 18, 2019, 11:48

@kdjstudios @jredfield 

Thank you for the update.  
We can release this tomorrow ( Tuesday). Right?

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @jredfield on Nov 18, 2019, 11:50

@kdjstudios Aavaz confirmed with me this morning that they have this all set to release tomorrow (Tuesday).  Their only hesitation was to wait until you had made any annoucements, newsletter notifications, etc.  If you say go ahead then they will.  I will remind them on WhatsApp before going to bed.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 18, 2019, 11:54

@amishra 

Yes please release this. We do not want any more billing cycles with issues.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 18, 2019, 22:22

@kdjstudios @jredfield 

Thank you for the confirmation.

We have successfully released on both of the production servers. 

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 19, 2019, 12:23

@amishra 

Thank you Marking this for confirmation to close.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 19, 2019, 12:24

I wrote:

> The support team found a discrepancy in the code which determines if a recurring fee (base rate) should be populated on invoices. They have since released an update to fix this. Please let us know if you have any questions or concerns, otherwise we will be marking this as closed.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 20, 2019, 10:31

closed
